plugins/optimization-detective/detect.js

```diff
@@ -361,46 +356,59 @@ export default async function detect( {
 			boundingClientRect: elementIntersection.boundingClientRect,
 		};
 
-		urlMetrics.elements.push( elementMetrics );
+		urlMetric.elements.push( elementMetrics );
 	}
 
 	if ( isDebug ) {
-		log( 'Current URL metrics:', urlMetrics );
+		log( 'Current URL metrics:', urlMetric );
 	}
 
-	// Yield to main before sending data to server to further break up task.
+	// Wait for the page to be hidden.
 	await new Promise( ( resolve ) => {
-		setTimeout( resolve, 0 );
-	} );
-
-	try {
-		const response = await fetch( restApiEndpoint, {
-			method: 'POST',
-			headers: {
-				'Content-Type': 'application/json',
-				'X-WP-Nonce': restApiNonce,
+		win.addEventListener( 'pagehide', resolve, { once: true } );
+		win.addEventListener( 'pageswap', resolve, { once: true } );
+		doc.addEventListener(
+			'visibilitychange',
+			() => {
+				if ( document.visibilityState === 'hidden' ) {
+					// TODO: This will fire even when switching tabs.
+					resolve();
+				}
 			},
-			body: JSON.stringify( urlMetrics ),
-		} );
+			{ once: true }
+		);
+	} );
 
-		if ( response.status === 200 ) {
-			setStorageLock( getCurrentTime() );
+	for ( const extension of extensions ) {
+		if ( extension.finalize instanceof Function ) {
+			extension.finalize( {
```

Comment:
So extensions add their own data by directly modifying the passed `urlMetric`? Why not have them return the new data instead?

Reply:
As this PR is currently implemented, Embed Optimizer overrides the `boundingClientRect` with the final rect reported the ResizeObserver for embeds. But since then I've realized this isn't ideal because we can't know if the `boundingClientRect` was the initial size or the final size. So in that case I think it would be better if the schema is made extensible (cf. https://github.com/WordPress/performance/pull/1490#issuecomment-2300905345) so that Embed Optimizer can add a `resizedBoundingClientRect` which we then exclusively look for when setting the `min-height`. So in that case, the extension's `finalize` function could just return an object that gets merged on top of the `urlMetric`. But it seems there could be use cases for extensions to still be able modify the entire object?

Reply:
I think returning an object and using `structuredClone` here to avoid unintended modifications is cleaner.

Reply:
OK, sounds good. But using `structuredClone()` where? We wouldn't need to pass in the `urlMetric` anymore to the `finalize` function, so would a clone be needed?

Reply:
Er wait, we do need to pass the `urlMetric` because Embed Optimizer needs to amend `Element` data in the `elements` key.

Reply:
Looking at this old comment thread, does it still require an update? Asking since `urlMetric` is still being passed below.

Reply:
I don't think it requires an update. The `urlMetric` object needs to be passed so that the extension can amend it as needed. Sure this would mean the extension could modify "core" properties, but I don't see this as a problem (and in fact it may be desirable).

Reply:
I think it's a fair concern that extensions should only be able to modify what we want them to modify. I don't think we have a good use-case in mind why someone _should_ modify "core" properties, so then we shouldn't allow it, until a real reasonable need arises.

Some alternatives for how to resolve this:
* Either don't give access to `urlMetric` at all and have the functions only return data that is centrally added to `urlMetric`. Downside: May make for a worse API, since returning e.g. just an object is not exactly safe (you would need to remember property names, there's no auto-completion etc.).
* We somehow "lock" things in the passed `urlMetric` so that they can't be written by the function. Not quite sure on how to reliably do that, but that could work.
* We pass another object to the function that only contains the bits of `urlMetric` that _should_ be amendable.

I think the last point would be a cleaner solution.

Reply:
The problem is that it's not just the `urlMetric` that needs to be modified. It's also the elements located in `urlMetric.elements`. The extensions need to be able to extend the properties of those elements. I'll think about an alternative.

Reply:
Work in progress: 7ca1fbc

Reply:
OK, I've created a new API whereby the `finalize` function is passed an object which includes functions:

```ts
export type FinalizeArgs = {
	readonly getRootData: () => URLMetric,
	readonly amendRootData: ( properties: AmendedRootData ) => void,
	readonly getElementData: ( xpath: string ) => ElementData|null,
	readonly amendElementData: ( xpath: string, properties: AmendedElementData ) => void,
	readonly isDebug: boolean,
};
```

If someone tries to set a core property of the `URLMetric` or on an `element`, then this triggers an error.

What do you think?

Reply:
@swissspidy Do you like?